PG.StarWarsGame.Files.MEG/PG.StarWarsGame.Files.MEG/Services/V1/MegFileProcessService.cs

```diff
@@ -0,0 +1,251 @@
+            }
+
+            uint headerSize = GetMegFileHeaderSize(filePath);
+            byte[] megFileHeader = new byte[headerSize];
+            using (BinaryReader reader = new BinaryReader(m_fileSystem.FileStream.Create(filePath, FileMode.Open)))
+            {
+                reader.Read(megFileHeader, 0, megFileHeader.Length);
+            }
+
+            MegFileBuilder builder = new MegFileBuilder();
+            MegFile megFile = builder.FromBytes(megFileHeader);
+            MegFileHolder holder = new MegFileHolder(m_fileSystem.Path.GetDirectoryName(filePath),
+                m_fileSystem.Path.GetFileNameWithoutExtension(filePath));
+            for (int i = 0; i < megFile.Header.NumFiles; i++)
+            {
+                string fileName = megFile.FileNameTable.MegFileNameTableRecords[i].FileName;
+                uint fileOffset = megFile.FileContentTable.MegFileContentTableRecords[i].FileStartOffsetInBytes;
+                uint fileSize = megFile.FileContentTable.MegFileContentTableRecords[i].FileSizeInBytes;
+                holder.Content.Add(new MegFileDataEntry(fileName, Convert.ToInt32(fileOffset), fileSize));
+            }
+
+            return holder;
+        }
+
+        private void CreateTargetDirectoryIfNotExists(string targetDirectory)
+        {
+            if (m_fileSystem.Directory.Exists(targetDirectory))
+            {
+                return;
+            }
+
+            m_logger?.LogWarning($"The given directory \"{targetDirectory}\" does not exist. Trying to create it.");
+            m_fileSystem.Directory.CreateDirectory(targetDirectory);
+        }
+
+        private void UnpackMegFilePreservingDirectoryHierarchy(MegFileHolder holder, string targetDirectory,
+            MegFileDataEntry megFileDataEntry)
+        {
+            string filePath = m_fileSystem.Path.Combine(targetDirectory, megFileDataEntry.RelativeFilePath);
+            string path = m_fileSystem.FileInfo.FromFileName(filePath).Directory.FullName;
+            CreateTargetDirectoryIfNotExists(path);
+            using BinaryReader reader = GetBinaryReaderForFileHolder(holder);
+            byte[] file = new byte[megFileDataEntry.Size];
+            reader.BaseStream.Seek(megFileDataEntry.Offset, SeekOrigin.Begin);
+            reader.Read(file, 0, file.Length);
+            m_fileSystem.File.WriteAllBytes(filePath, file);
+        }
+
+        private void UnpackMegFileFlatDirectoryHierarchy(MegFileHolder holder, string targetDirectory,
+            MegFileDataEntry megFileDataEntry)
+        {
+            string filePath = m_fileSystem.Path.Combine(targetDirectory, megFileDataEntry.RelativeFilePath);
+            filePath = m_fileSystem.Path.Combine(targetDirectory, m_fileSystem.Path.GetFileName(filePath));
+            using BinaryReader reader = GetBinaryReaderForFileHolder(holder);
+            byte[] file = new byte[megFileDataEntry.Size];
```

Comment:
This should probably be moved to a fully buffered approach similar to the packing logic.

Reply:
indeed